### PR TITLE
fix tmy3 leapyear in February handling, coerce_year raises exception

### LINKED ISSUE #865: tmy with coerce year fails if February contains leap year, eg Greensboro-NC
**Describe the bug**
Try to read [Greensboro TMY3](https://rredc.nrel.gov/solar/old_data/nsrdb/1991-2005/data/tmy3/723170TYA.CSV) with `coerce_year=1990`

raises the follow traceback:
<details>

```python
In [1]: import pvlib

In [2]: greensboro = pvlib.iotools.read_tmy3('723170TYA.CSV', coerce_year=1990)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pandas\io\parsers.py in converter(*date_cols)
   3280                 result = tools.to_datetime(
-> 3281                     date_parser(*date_cols), errors="ignore", cache=cache_dates
   3282                 )

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pvlib\iotools\tmy.py in <lambda>(*x)
    194         parse_dates={'datetime': ['Date (MM/DD/YYYY)', 'Time (HH:MM)']},
--> 195         date_parser=lambda *x: _parsedate(*x, year=coerce_year),
    196         index_col='datetime')

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pvlib\iotools\tmy.py in _parsedate(ymd, hour, year)
    215     # and dateutil's inability to handle that.
--> 216     offset_hour = int(hour[:2]) - 1
    217     offset_datetime = '{} {}:00'.format(ymd, offset_hour)

TypeError: only size-1 arrays can be converted to Python scalars

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pandas\io\parsers.py in converter(*date_cols)
   3291                             parser=date_parser,
-> 3292                             dayfirst=dayfirst,
   3293                         ),

pandas/_libs/tslibs/parsing.pyx in pandas._libs.tslibs.parsing.try_parse_dates()

pandas/_libs/tslibs/parsing.pyx in pandas._libs.tslibs.parsing.try_parse_dates()

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pvlib\iotools\tmy.py in <lambda>(*x)
    194         parse_dates={'datetime': ['Date (MM/DD/YYYY)', 'Time (HH:MM)']},
--> 195         date_parser=lambda *x: _parsedate(*x, year=coerce_year),
    196         index_col='datetime')

TypeError: _parsedate() missing 1 required positional argument: 'hour'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-2-6078abbb94dd> in <module>
----> 1 greensboro = pvlib.iotools.read_tmy3('723170TYA.CSV', coerce_year=1990)

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pvlib\iotools\tmy.py in read_tmy3(filename, coerce_year, recolumn)
    194         parse_dates={'datetime': ['Date (MM/DD/YYYY)', 'Time (HH:MM)']},
    195         date_parser=lambda *x: _parsedate(*x, year=coerce_year),
--> 196         index_col='datetime')
    197
    198     if recolumn:

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pandas\io\parsers.py in parser_f(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)
    683         )
    684
--> 685         return _read(filepath_or_buffer, kwds)
    686
    687     parser_f.__name__ = name

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pandas\io\parsers.py in _read(filepath_or_buffer, kwds)
    461
    462     try:
--> 463         data = parser.read(nrows)
    464     finally:
    465         parser.close()

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pandas\io\parsers.py in read(self, nrows)
   1152     def read(self, nrows=None):
   1153         nrows = _validate_integer("nrows", nrows)
-> 1154         ret = self._engine.read(nrows)
   1155
   1156         # May alter columns / col_dict

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pandas\io\parsers.py in read(self, nrows)
   2132             data = {k: v for k, (i, v) in zip(names, data)}
   2133
-> 2134             names, data = self._do_date_conversions(names, data)
   2135             index, names = self._make_index(data, alldata, names)
   2136

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pandas\io\parsers.py in _do_date_conversions(self, names, data)
   1883                 self.index_names,
   1884                 names,
-> 1885                 keep_date_col=self.keep_date_col,
   1886             )
   1887

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pandas\io\parsers.py in _process_date_conversion(data_dict, converter, parse_spec, index_col, index_names, columns, keep_date_col)
   3355
   3356             _, col, old_names = _try_convert_dates(
-> 3357                 converter, colspec, data_dict, orig_names
   3358             )
   3359

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pandas\io\parsers.py in _try_convert_dates(parser, colspec, data_dict, columns)
   3388     to_parse = [data_dict[c] for c in colnames if c in data_dict]
   3389
-> 3390     new_col = parser(*to_parse)
   3391     return new_name, new_col, colnames
   3392

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pandas\io\parsers.py in converter(*date_cols)
   3295                     )
   3296                 except Exception:
-> 3297                     return generic_parser(date_parser, *date_cols)
   3298
   3299     return converter

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pandas\io\date_converters.py in generic_parser(parse_func, *cols)
     36     for i in range(N):
     37         args = [c[i] for c in cols]
---> 38         results[i] = parse_func(*args)
     39
     40     return results

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pvlib\iotools\tmy.py in <lambda>(*x)
    193         csvdata, header=0,
    194         parse_dates={'datetime': ['Date (MM/DD/YYYY)', 'Time (HH:MM)']},
--> 195         date_parser=lambda *x: _parsedate(*x, year=coerce_year),
    196         index_col='datetime')
    197

~\AppData\Local\Continuum\miniconda3\envs\py37\lib\site-packages\pvlib\iotools\tmy.py in _parsedate(ymd, hour, year)
    219     true_date = offset_date + dateutil.relativedelta.relativedelta(hours=1)
    220     if year is not None:
--> 221         true_date = true_date.replace(year=year)
    222     return true_date
    223

ValueError: day is out of range for month
```
</details>

**To Reproduce**
1. download tmy3 for greensboro
2. start python import pvlib
3. execute `pvlib.iotools.read_tmy3('723170TYA.CSV', coerce_year=1990)`

Also, _seems_ to works fine if `coerce_year=None`, but actually raises a different silent issue:
image
**Note the February 29th at midnite in between Feb. 28th and March 1st**

**Versions:**
 - ``pvlib.__version__``: 0.7.1
 - ``pandas.__version__``: 0.25.3
 - python: 3.7.6
- numpy: 1.18.1

**Additional context**
Look at cells A1416 around end of February:

2/28/1996 | 22:00
-- | --
2/28/1996 | 23:00
2/28/1996 | 24:00:00
3/1/1990 | 1:00
3/1/1990 | 2:00
3/1/1990 | 3:00

PS you can paste from Excel into GitHub and it GH automatically formats the data as MD table, nice!

### PR DESCRIPTION
**Pull request by mikofski**
- [x] Closes #865
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

check if the parsed true date is February 29th, and if so, reset the non-leap year day to March 1st (`nlyearday=60`)

### CODE REVIEW

File: pvlib/tests/iotools/test_tmy.py [hunk 1]
```diff
@@ -41,3 +45,37 @@ def test_read_tmy3_no_coerce_year():
...
+        'State': 'NC',
+        'TZ': -5.0,
+        'latitude': 36.1,
+        'longitude': -79.95,
+        'altitude': 273.0}
+    assert meta == greensboro
+    # February for Greensboro is 1996, a leap year, so check to make sure there
+    # aren't any rows in the output that contain Feb 29th
+    assert data['1996-02-29 00:00'].size == 0
+    # just to be safe, make sure this _IS_ the Greensboro file
```
Comment (CameronTStark) [inline pvlib/tests/iotools/test_tmy.py]: This appears to be duplicated code.

**Comment by CameronTStark** [comment]
Good catch @mikofski, thanks for submitting this!

**Comment by CameronTStark** [comment]
The macOS_Mojave tests had an error "requests.exceptions.HTTPError: Error Code,Error Message "OVER_RATE_LIMIT"" which appears due to a request to NREL's API. I'm not familiar but could this be due to a monthly quota that we're exceeding with our CI?

**Comment by mikofski** [comment]
>The macOS_Mojave tests had an error "requests.exceptions.HTTPError: Error Code,Error Message "OVER_RATE_LIMIT"" which appears due to a request to NREL's API. I'm not familiar but could this be due to a monthly quota that we're exceeding with our CI?

yes, I think it's due to the [NREL PSM3 demo key throttle limits](https://developer.nrel.gov/docs/rate-limits/)  - if it becomes a bottle neck to testing we can set up an [NREL API key](https://developer.nrel.gov/) and save it in an [Azure secret variable](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#secret-variables) to use just for testing, so that no secrets are published to git

**Comment by kanderso-nrel** [comment]
I haven't tested it thoroughly, but would this work?  Seems more straightforward (to my taste anyway) and I'd guess that it's faster than using a custom date parser function:

```
data = pd.read_csv(csvdata, header=0, index_col='Date (MM/DD/YYYY)')
data.index = pd.to_datetime(data.index, format='%m/%d/%Y')
shifted_hour = data['Time (HH:MM)'].str[:2].astype(int) - 1
data.index += pd.to_timedelta(shifted_hour, unit='hour')
if coerce_year is not None:
    data.index = data.index.map(lambda dt: dt.replace(year=coerce_year))
```

**Comment by mikofski** [comment]
@kanderso-nrel that looks much cleaner, here's what I get:

## current implementation

```python
In [2]: %timeit greensboro = pvlib.iotools.read_tmy3('723170TYA.CSV', 1990)
780 ms ± 9.14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [3]: %timeit greensboro = pvlib.iotools.read_tmy3('723170TYA.CSV')
759 ms ± 5.91 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

## proposed implementation

#### testing

Yes, it does indeed pass pytest

<details>

```
python c:\Users\mikm\.vscode\extensions\ms-python.python-2020.1.58038\pythonFiles\testing_tools\run_adapter.py discover pytest -- --rootdir c:\Users\mikm\Projects\pvlib-python -s --cache-clear
============================= test session starts =============================
platform win32 -- Python 3.6.6, pytest-5.3.1, py-1.7.0, pluggy-0.13.1
rootdir: c:\Users\mikm\Projects\pvlib-python, inifile: ..\pytest.ini
plugins: cov-2.8.1, mock-1.12.1, timeout-1.3.3
collected 8 items

pvlib\tests\iotools\test_tmy.py ........                                 [100%]

============================== warnings summary ===============================
venv\lib\site-packages\_pytest\junitxml.py:436
  c:\Users\mikm\Projects\pvlib-python\venv\lib\site-packages\_pytest\junitxml.py:436: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0.
  Add 'junit_family=legacy' to your pytest.ini file to silence this warning and make your suite compatible.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
 generated xml file: C:\Users\mikm\AppData\Local\Temp\tmp-3216t5w06AW1b4VE.xml 
======================== 8 passed, 1 warning in 6.20s =========================
```

</details>

And it _is_ about 5x to 7x faster than current implementation, and it removes `datutil` import

```python
In [1]: import pvlib

In [2]: %timeit greensboro = pvlib.iotools.read_tmy3('723170TYA.CSV', 1990)
151 ms ± 5.58 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [3]: %timeit greensboro = pvlib.iotools.read_tmy3('723170TYA.CSV')
92.4 ms ± 3.09 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by kanderso-nrel** [comment]
I think the failing tests are because `unit='hour'` for `pd.to_timedelta` was added in pandas 0.24.0 and they're using 0.23.4.  Switching to `unit='h'` should be safe?  Sorry for not being more thorough the first time. 

- https://pandas.pydata.org/pandas-docs/stable/whatsnew/v0.24.0.html#timedelta
- https://github.com/pandas-dev/pandas/pull/23439/files#diff-362279ed6b3644d87b20017b92db3433R27

**Comment by mikofski** [comment]
OK, but [Travis-ci is using pandas-0.18.1](https://travis-ci.org/pvlib/pvlib-python/jobs/641462752#L457) which is the currently published pvlib-0.7.1 minimum requirement as of [v0.6.3](https://pvlib-python.readthedocs.io/en/latest/whatsnew.html#v0-7-0-december-18-2019), although it looks like the [unit string 'h' also worked for pandas-0.18.1] so all good - added a NOTE if we ever bump pandas that this constraint could be relaxed if needed, although code change might not be necessary.

**Comment by mikofski** [comment]
pls ignore rtd build errors, they are sporadic 😕

File: pvlib/iotools/tmy.py [hunk 2]
```diff
@@ -140,6 +139,9 @@ def read_tmy3(filename=None, coerce_year=None, recolumn=True):
     TMYData.PresWthUncertainty          Present weather code uncertainty, see [2]_.
     =============================       ======================================================================================================================================================
 
+    .. warning:: when coercing the year, the last index in the dataframe will
+        actually be the first hour of the year
```
Comment (cwhanse) [inline pvlib/iotools/tmy.py]:
```suggestion
        actually be the first hour of the following year. TMY3 uses hour-ending timestamps, and the timestamp for the hour 23:00-24:00 is 24:00. pandas converts 24:00 to 00:00 of the following day.
```

Comment (cwhanse) [reply pvlib/iotools/tmy.py]: A bit more explanation may help some users.

**Comment by mikofski** [comment]
Hi Cliff, here's what I've found:

* If `coerce_year is None` then as you suggested, the last hour will be the first day of the next year. _EG_:  if TMY3 was `1988-12-31 24:00:00` this becomes `1989-01-01 00:00:00` 

* _But_ if `coerce_year` is set to an art year, say 1990, then what I meant was that, the last hour will be the first day of the same year. _EG_:  if TMY3 was `1988-12-31 24:00:00` this becomes `1990-01-01 00:00:00` which could be surprising to some users

Should we improve the documentation warning to elaborate on both scenarios?

**Comment by cwhanse** [comment]
> But if coerce_year is set to an art year, say 1990, then what I meant was that, the last hour will be the first day of the same year. EG:  if TMY3 was 1988-12-31 24:00:00 this becomes 1990-01-01 00:00:00 which could be surprising to some users

So the output would start with 1900-01-01 01:00:00 _and_ end with 1900-01-01 00:00:00?

**Comment by kanderso-nrel** [comment]
I don't think the updated code suffers from this "year-rollover" issue, does it?  I think regardless of whether the year is coerced, the first and last index values will be `XXXX-01-01 00:00:00` and `YYYY-12-31 23:00:00`.

**Comment by cwhanse** [comment]
> I don't think the updated code suffers from this "year-rollover" issue, does it?  I think regardless of whether the year is coerced, the first and last index values will be XXXX-01-01 00:00:00 and YYYY-12-31 23:00:00.

If this is the case, then we should inform the user that pvlib applies an hour-beginning convention, since TMY3 uses an hour-ending timestamp convention.

**Comment by mikofski** [comment]
😞 Ugh, Kevin, you are right, the timestamps have now been shifted to the beginning of the hour ...
<details>

```
In [1]: import pvlib

In [2]: greensboro1990 = pvlib.iotools.read_tmy3('723170TYA.CSV', 1990)

In [3]: greensboro = pvlib.iotools.read_tmy3('723170TYA.CSV')

In [5]: greensboro1990[0][['GHI', 'DNI', 'DHI']].head()
Out[5]: 
                           GHI  DNI  DHI
1990-01-01 00:00:00-05:00    0    0    0
1990-01-01 01:00:00-05:00    0    0    0
1990-01-01 02:00:00-05:00    0    0    0
1990-01-01 03:00:00-05:00    0    0    0
1990-01-01 04:00:00-05:00    0    0    0

In [6]: greensboro1990[0][['GHI', 'DNI', 'DHI']].tail()
Out[6]: 
                           GHI  DNI  DHI
1990-12-31 19:00:00-05:00    0    0    0
1990-12-31 20:00:00-05:00    0    0    0
1990-12-31 21:00:00-05:00    0    0    0
1990-12-31 22:00:00-05:00    0    0    0
1990-12-31 23:00:00-05:00    0    0    0

In [7]: greensboro[0][['GHI', 'DNI', 'DHI']].head()
Out[7]: 
                           GHI  DNI  DHI
1988-01-01 00:00:00-05:00    0    0    0
1988-01-01 01:00:00-05:00    0    0    0
1988-01-01 02:00:00-05:00    0    0    0
1988-01-01 03:00:00-05:00    0    0    0
1988-01-01 04:00:00-05:00    0    0    0

In [8]: greensboro[0][['GHI', 'DNI', 'DHI']].tail()
Out[8]: 
                           GHI  DNI  DHI
1980-12-31 19:00:00-05:00    0    0    0
1980-12-31 20:00:00-05:00    0    0    0
1980-12-31 21:00:00-05:00    0    0    0
1980-12-31 22:00:00-05:00    0    0    0
1980-12-31 23:00:00-05:00    0    0    0
```
</details>

which IIRC contradicts of the TMY3 manual, _EG_ for GHI:
>Direct and diffuse horizontal radiation recv’d during 60 minutes prior to timestamp, Wh/m^2

And differs from how pvlib currently parses TMY3:
<details>

```
In [1]: import pvlib

In [9]: pvlib.__version__
Out[9]: '0.7.1'

In [2]: hickory1990 = pvlib.iotools.read_tmy3('723145TYA.CSV', 1990)

In [3]: hickory = pvlib.iotools.read_tmy3('723145TYA.CSV')

In [5]: hickory1990[0][['GHI','DNI','DHI']].head()
Out[5]:
                           GHI  DNI  DHI
datetime
1990-01-01 01:00:00-05:00    0    0    0
1990-01-01 02:00:00-05:00    0    0    0
1990-01-01 03:00:00-05:00    0    0    0
1990-01-01 04:00:00-05:00    0    0    0
1990-01-01 05:00:00-05:00    0    0    0

In [6]: hickory1990[0][['GHI','DNI','DHI']].tail()
Out[6]:
                           GHI  DNI  DHI
datetime
1990-12-31 20:00:00-05:00    0    0    0
1990-12-31 21:00:00-05:00    0    0    0
1990-12-31 22:00:00-05:00    0    0    0
1990-12-31 23:00:00-05:00    0    0    0
1990-01-01 00:00:00-05:00    0    0    0

In [7]: hickory[0][['GHI','DNI','DHI']].head()
Out[7]:
                           GHI  DNI  DHI
datetime
2004-01-01 01:00:00-05:00    0    0    0
2004-01-01 02:00:00-05:00    0    0    0
2004-01-01 03:00:00-05:00    0    0    0
2004-01-01 04:00:00-05:00    0    0    0
2004-01-01 05:00:00-05:00    0    0    0

In [8]: hickory[0][['GHI','DNI','DHI']].tail()
Out[8]:
                           GHI  DNI  DHI
datetime
1999-12-31 20:00:00-05:00    0    0    0
1999-12-31 21:00:00-05:00    0    0    0
1999-12-31 22:00:00-05:00    0    0    0
1999-12-31 23:00:00-05:00    0    0    0
2000-01-01 00:00:00-05:00    0    0    0

```
</details>

so IMO we would need to fix this before merging, because the way that reads now is that for example on Jan1 at 8AM, the preceding hour had a cummulative GHI of 46[Wh/m^2]
<details>

```
In [11]: greensboro1990[0][['GHI', 'DNI', 'DHI']].loc['1990-01-01 06:00':'1990-01-01 18:00']
Out[11]: 
                           GHI  DNI  DHI
1990-01-01 06:00:00-05:00    0    0    0
1990-01-01 07:00:00-05:00    9    1    9
1990-01-01 08:00:00-05:00   46    3   46
1990-01-01 09:00:00-05:00   79    4   78
1990-01-01 10:00:00-05:00  199    3  198
1990-01-01 11:00:00-05:00  261    3  260
1990-01-01 12:00:00-05:00  155    0  155
1990-01-01 13:00:00-05:00  144    2  144
1990-01-01 14:00:00-05:00  131    1  131
1990-01-01 15:00:00-05:00   81    1   81
1990-01-01 16:00:00-05:00   49    1   49
1990-01-01 17:00:00-05:00    4    0    4
1990-01-01 18:00:00-05:00    0    0    0
```
</details>

instead of at 9AM


Date (MM/DD/YYYY) | Time (HH:MM) | ETR (W/m^2) | ETRN (W/m^2) | GHI (W/m^2) | GHI source | GHI uncert (%) | DNI (W/m^2) | DNI source | DNI uncert (%) | DHI (W/m^2)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1/1/1988 | 6:00 | 0 | 0 | 0 | 1 | 0 | 0 | 1 | 0 | 0
1/1/1988 | 7:00 | 0 | 0 | 0 | 1 | 0 | 0 | 1 | 0 | 0
1/1/1988 | 8:00 | 25 | 649 | 9 | 1 | 13 | 1 | 1 | 9 | 9
1/1/1988 | 9:00 | 228 | 1415 | 46 | 1 | 13 | 3 | 1 | 9 | 46
1/1/1988 | 10:00 | 439 | 1415 | 79 | 1 | 9 | 4 | 1 | 9 | 78
1/1/1988 | 11:00 | 598 | 1415 | 199 | 1 | 9 | 3 | 1 | 9 | 198
1/1/1988 | 12:00 | 696 | 1415 | 261 | 1 | 9 | 3 | 1 | 9 | 260
1/1/1988 | 13:00 | 723 | 1415 | 155 | 1 | 9 | 0 | 1 | 9 | 155
1/1/1988 | 14:00 | 680 | 1415 | 144 | 1 | 9 | 2 | 1 | 9 | 144
1/1/1988 | 15:00 | 568 | 1415 | 131 | 1 | 9 | 1 | 1 | 9 | 131
1/1/1988 | 16:00 | 395 | 1415 | 81 | 1 | 9 | 1 | 1 | 9 | 81
1/1/1988 | 17:00 | 175 | 1415 | 49 | 1 | 13 | 1 | 1 | 9 | 49
1/1/1988 | 18:00 | 6 | 342 | 4 | 1 | 13 | 0 | 1 | 9 | 4

**Comment by cwhanse** [comment]
> 😞 Ugh, Kevin, you are right, the timestamps have now been shifted to the beginning of the hour ...
> which IIRC contradicts of the TMY3 manual, EG for GHI:

I think it is wise to stick with the TMY3 time convention (hour-ending), so as not to surprise users who are used to that.

**Comment by kanderso-nrel** [comment]
Fair enough -- in that case, what should be done about the timestamp `YYYY-02-28 24:00` when coercing to a leap year?  I think to be consistent with hour-end timestamps, this is the correct output:
```
...
2000-02-28 22:00:00-05:00        22:00    0  ...              C                   8
2000-02-28 23:00:00-05:00        23:00    0  ...              C                   8
2000-02-29 00:00:00-05:00        24:00    0  ...              C                   8
2000-03-01 01:00:00-05:00        01:00    0  ...              C                   8
2000-03-01 02:00:00-05:00        02:00    0  ...              C                   8
...
```

Which can be achieved by removing the `- 1` when calculating the shifted_hour and swapping the order of the year coercion and adding the hourly timedeltas (coerce first so that feb29 exists to shift into).

**Comment by mikofski** [comment]
So I made some changes rather than reverting back to the previous code, and it is still faster, and now seems to be working as it did before, although now the `Date(MM/DD/YYYY)` and Time(HH:MM)` columns are still in dataframe which I kind of like so the user can actually see what the index was derived from, okay

## new metrics
<details>

```
In [1]: import pvlib

In [2]: %timeit greensboro = pvlib.iotools.read_tmy3('723170TYA.CSV')
117 ms ± 1.61 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [3]: %timeit greensboro1990 = pvlib.iotools.read_tmy3('723170TYA.CSV', 1990)
190 ms ± 4.91 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: greensboro = pvlib.iotools.read_tmy3('723170TYA.CSV')

In [5]: greensboro[0][['GHI', 'DNI', 'DHI']].tail()
Out[5]: 
                           GHI  DNI  DHI
1980-12-31 20:00:00-05:00    0    0    0
1980-12-31 21:00:00-05:00    0    0    0
1980-12-31 22:00:00-05:00    0    0    0
1980-12-31 23:00:00-05:00    0    0    0
1981-01-01 00:00:00-05:00    0    0    0

In [6]: greensboro[0][['GHI', 'DNI', 'DHI']].head()
Out[6]: 
                           GHI  DNI  DHI
1988-01-01 01:00:00-05:00    0    0    0
1988-01-01 02:00:00-05:00    0    0    0
1988-01-01 03:00:00-05:00    0    0    0
1988-01-01 04:00:00-05:00    0    0    0
1988-01-01 05:00:00-05:00    0    0    0

In [8]: greensboro1990 = pvlib.iotools.read_tmy3('723170TYA.CSV', 1990)

In [9]: greensboro1990[0][['GHI', 'DNI', 'DHI']].tail()
Out[9]: 
                           GHI  DNI  DHI
1990-12-31 20:00:00-05:00    0    0    0
1990-12-31 21:00:00-05:00    0    0    0
1990-12-31 22:00:00-05:00    0    0    0
1990-12-31 23:00:00-05:00    0    0    0
1990-01-01 00:00:00-05:00    0    0    0

In [10]: greensboro1990[0][['GHI', 'DNI', 'DHI']].head()
Out[10]: 
                           GHI  DNI  DHI
1990-01-01 01:00:00-05:00    0    0    0
1990-01-01 02:00:00-05:00    0    0    0
1990-01-01 03:00:00-05:00    0    0    0
1990-01-01 04:00:00-05:00    0    0    0
1990-01-01 05:00:00-05:00    0    0    0
```
</details>

**Comment by wholmgren** [comment]
This PR is about fixing the parser when the TMY file uses February data from a leap year rather than what year we're coercing into, right? I see that your test file excludes the leap day and I understand that this causes problems because the labels are hour ending and pandas uses `00:00` for midnight instead of `24:00`.

Perhaps a separate issue, but would you want to duplicate Feb 28 into Feb 29 when coercing into a leap year? Annual yield for a leap year should be ~1/365th larger than for a non-leap year. I would lean towards no, but thought it was worth mentioning.

**Comment by mikofski** [comment]
I think that the current code is a little better, passes the tests, and I've added tests to check for what we've done

>I think to be consistent with hour-end timestamps, this is the correct output ...

If it's okay with everyone, I have a preference for TMY files to never have a leapday in it

>Perhaps a separate issue, but would you want to duplicate Feb 28 into Feb 29 when coercing into a leap year? Annual yield for a leap year should be ~1/365th larger than for a non-leap year. I would lean towards no, but thought it was worth mentioning.

IIUC we are both agreeing that a TMY file should only have 8760 hours, and neither contain an extra day nor should it have the `YYYY-02-29 00:00` please correct me if I'm wrong


FYI: the test CI and RTD failures are unrelated, Travis failed on forecast HRRR, bare linux couldn't reach MIDC or TMY3 remote servers, macOS failed because we hit the psm3 throttle limit

**Comment by cwhanse** [comment]
> If it's okay with everyone, I have a preference for TMY files to never have a leapday in it

+1

> IIUC we are both agreeing that a TMY file should only have 8760 hours, and neither contain an extra day nor should it have the YYYY-02-29 00:00 please correct me if I'm wrong

I think it would be courteous to document this behavior, and perhaps warn if a leap year is requested.

> macOS failed because we hit the psm3 throttle limit

Could it be that users are supplying the pvlib demo key to the NSRDB? We frequently hit this throttle for testing, and I don't think we do that much unit tests.   Separate issue.

File: pvlib/tests/iotools/test_tmy.py [hunk 3]
```diff
@@ -41,3 +45,31 @@ def test_read_tmy3_no_coerce_year():
...
+        'Name': '"GREENSBORO PIEDMONT TRIAD INT"',
+        'State': 'NC',
+        'TZ': -5.0,
+        'latitude': 36.1,
+        'longitude': -79.95,
+        'altitude': 273.0}
+    assert meta == greensboro
+    # February for Greensboro is 1996, a leap year, so check to make sure there
+    # aren't any rows in the output that contain Feb 29th
+    assert data['1996-02-29 00:00'].size == 0
```
Comment (wholmgren) [inline pvlib/tests/iotools/test_tmy.py]: It's hard for me to track what's actually supposed to happen. Perhaps a test like

```python
assert data.index[67] = pd.Timestamp('1996-02-28 23:00')
assert data.index[68] = pd.Timestamp('1996-03-01 00:00')
```

assuming that's actually correct. And similar for the coerced year.

**Comment by mikofski** [comment]
> I think it would be courteous to document this behavior, and perhaps warn if a leap year is requested.

good idea, IIUC then something like this, to [check if `coerce_year` is a leap year](https://docs.python.org/3/library/calendar.html#calendar.isleap):
```python
import calendar
import warnings

# in read_tmy3()
if calendar.isleap(coerce_year):
    warnings.warn(f'{coerce_year} is a leap year, but the TMY data will not contain 2/29.')
```

If it's okay, can we move this to a new, separate issue/feature? I think this may require a little discussion, and I may need pause working on this PR for a little while.


>Could it be that users are supplying the pvlib demo key to the NSRDB? We frequently hit this throttle for testing, and I don't think we do that much unit tests. Separate issue.

Agreed, see my [comment above](#issuecomment-578202317) - I think we should make a separate issue to get a dedicated test API key from NREL, and then upload it securely to the CI's as secret variables, and make sure that the API keys are not accidentally published to the git repo

**Comment by mikofski** [comment]
Now I am getting [this error from Travis py35-min CI](https://travis-ci.org/pvlib/pvlib-python/jobs/641587545#L694):
    ValueError: cannot reindex from a duplicate axis

on this line
https://github.com/pvlib/pvlib-python/blob/d49b53a03cc3080793203499b12c454ebc9bafcc/pvlib/iotools/tmy.py#L221

@kanderso-nrel any help? The only difference is that now I use `pd.DatetimeIndex` and before the index was created by `read_csv`

**Comment by wholmgren** [comment]
I agree with mark’s concerns about scope creep. Looks good to me. Ok to merge assuming test failures are still unrelated.

**Comment by mikofski** [comment]
>Ok to merge assuming test failures are still unrelated.

still one error trying to fix, 😕 see if my last commit does it

**Comment by mikofski** [comment]
YES!!!! :tada: okay it's finally ready 🚀

**Comment by mikofski** [comment]
wait, I'm so stoopid, that line isn't even necessary! ha!

**Comment by mikofski** [comment]
OK to merge, thanks all 😄 !

**Comment by mikofski** [comment]
@CameronTStark ok to merge?

**Pull request merged**